### PR TITLE
Refactor/simplify extraction.ts

### PR DESCRIPTION
**Pull request by mcmire**
## Explanation

The goal of this commit is to make `extraction.ts` easier to follow. Since we are working with the TypeScript AST, we have to accept that the code will be somewhat dense, and we will need to ask maintainers to do some amount of studying to develop a mental model in their head. However, the more we can do to assist this process, the better.

To this end, this commit:

- Adds comments to illustrate steps in the parsing process and underlying nodes in the AST that are being visited
- Renames some functions to more accurately describe their contents
- Removes unnecessary `istanbul ignore` comments in favor of either adding tests that exercise the underlying logic or using assumptions (non-null assertions).
- Rewrite some JSDoc so it's less technical and more natural sounding

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [ ] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've communicated my changes to consumers by [updating changelogs for packages I've changed](https://github.com/MetaMask/core/tree/main/docs/processes/updating-changelogs.md)
- [ ] I've introduced [breaking changes](https://github.com/MetaMask/core/tree/main/docs/processes/breaking-changes.md) in this PR and have prepared draft pull requests for clients and consumer packages to resolve them

---

> [!NOTE]
> **Medium Risk**
> Changes core AST extraction logic that drives published API docs; behavior should stay equivalent but edge-case handling shifted to type-checker resolution and explicit skips.
> 
> **Overview**
> Refactors **`platform-api-docs`** messenger extraction for readability and safer edge-case handling, without changing the public doc-generation pipeline’s overall shape.
> 
> **`extraction.ts`** is reorganized into clearer phases: find `*Messenger` type aliases, recursively walk Actions/Events type parameters to collect capability declarations, then extract **`MessengerCapabilityPacket`** records (renamed from `ExtractedMessengerCapabilityType`) via inline object literals or `ControllerGetStateAction` / `ControllerStateChangeEvent` constructors. Discovery and extraction helpers are renamed and documented with AST-oriented comments; JSDoc handling is tightened (e.g. skip nameless `@param` tags, bare `@deprecated`).
> 
> **Type resolution** for capability `type` strings and constructor namespace args now leans on the type checker (`getType().isStringLiteral()`) instead of dedicated template-literal / `typeof` helpers. Invalid shapes are skipped explicitly (missing `type`/`handler`/`payload`, numeric `type`, qualified `Ns.Type` references, non–string-literal constructor args, method signatures mixed into object types).
> 
> **Tests** in `extraction.test.ts` add broad coverage for those skip/fallback paths; several `istanbul ignore` branches are removed in favor of tests or compile-time assumptions.
> 
> **`generate.ts`**, **`markdown.ts`**, and **`types.ts`** only adopt the new type name.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 57ad0bd3313638b45aab90ecaa836a452d977bc9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/platform-api-docs/src/extraction.ts [hunk 1]
```diff
@@ -110,25 +114,23 @@ function extractJsDoc(node: JSDocableNode): {
...
-      // istanbul ignore next: bare `@deprecated` (without explanatory text)
-      // doesn't appear in messenger JSDoc in practice.
       deprecatedLines.push(
         comment ? `**Deprecated:** ${comment}` : '**Deprecated:**',
       );
     } else if (tagName === 'param' && NodeGuards.isJSDocParameterTag(tag)) {
       const nameNode = tag.getNameNode();
-      // istanbul ignore next: `@param` tags without a name aren't valid JSDoc.
-      if (!nameNode) {
+      const paramName = nameNode.getText();
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:122]: It's true that if a `@param` tag does not have a name, it will be flagged by the ESLint JSDoc lint rules. However, `ts-morph` does not care about that. We instead explicitly check for this kind of tag and ignore it.

File: packages/platform-api-docs/src/extraction.ts [hunk 2]
```diff
@@ -158,54 +160,6 @@ function isDeprecated(node: JSDocableNode): boolean {
...
- * or a string literal.
- *
- * @param typeArg - The first generic argument node.
- * @returns The resolved namespace, or null.
- */
-function resolveNamespaceFromTypeArg(typeArg: TsMorphNode): string | null {
-  if (NodeGuards.isTypeQuery(typeArg)) {
-    const type = typeArg.getType();
-    if (type.isStringLiteral()) {
-      return type.getLiteralValueOrThrow() as string;
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:194]: We don't need to have various strategies of resolving the value of a string literal. The `getLiteralValueOrThrow` method on the type does this already. So we can simplify this.

File: packages/platform-api-docs/src/extraction.ts [hunk 3]
```diff
@@ -322,268 +272,385 @@ function buildMethodInfo(method: MethodDeclaration): MethodInfo {
...
+ * Looks for Messenger types in the source file (that is, those that are type
+ * aliases whose names end with "Messenger"), then extracts the `Actions` and
+ * `Events` parameters from these types.
  *
  * @param sourceFile - The TypeScript source file to scan.
- * @returns The list of locally- and transitively-referenced capability types.
+ * @returns A list of objects that represent messenger types.
  */
-function collectMessengerCapabilityTypeDeclarations(
+function findMessengerTypeAliases(
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:300]: I split this function off from `collectMessengerCapabilityTypeDeclarations` because it seemed like a logical step in the process that deserved its own name.

File: packages/platform-api-docs/src/extraction.ts [hunk 4]
```diff
@@ -322,268 +272,385 @@ function buildMethodInfo(method: MethodDeclaration): MethodInfo {
...
- *
- * walking the Actions slot yields each individual `NetworkController*Action`
- * declaration — both the local `NetworkControllerGetStateAction` and the
- * cross-file ones reached via `NetworkControllerMethodActions`.
+ * @param parsedMessengerTypeAliases - The list of objects representing
+ * messenger types, parsed in a previous step.
+ * @returns The list of type aliases that represent messenger capabilities among
+ * the given messenger types.
+ */
+function findAllMessengerCapabilityTypeDeclarations(
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:342]: I renamed this from `collectMessengerCapabilityTypeDeclarations` to `findAllMessengerCapabilityTypeDeclarations`. This was mainly to match `recursivelyFindMessengerCapabilityTypeDeclarations`.

File: packages/platform-api-docs/src/extraction.ts [hunk 5]
```diff
@@ -322,268 +272,385 @@ function buildMethodInfo(method: MethodDeclaration): MethodInfo {
...
- * duplicate work).
+ * @param node - The `ts-morph` AST node to walk.
+ * @param kind - Whether to tag found type aliases as 'action' or 'event'.
+ * @param visitedTypeDeclarations - A variable that tracks visited type aliases
+ * and prevents duplicates.
+ * @returns The list of extracted messenger capability type aliases as well as
+ * an updated version of `visitedTypeDeclarations`.
  */
-function walkCapabilityTypes(
+function recursivelyFindMessengerCapabilityTypeDeclarations(
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:384]: I renamed `walkCapabilityTypes` to `recursivelyFindMessengerCapabilityTypeDeclarations` as I felt like "walk" didn't accurately convey what we are trying to do. We are not merely visiting nodes, we are also capturing the capability types.

File: packages/platform-api-docs/src/extraction.ts [hunk 6]
```diff
@@ -322,268 +272,385 @@ function buildMethodInfo(method: MethodDeclaration): MethodInfo {
...
   if (!NodeGuards.isTypeReference(node)) {
-    return;
+    return result;
   }
+
   const nameNode = node.getTypeName();
-  // istanbul ignore next: qualified-name references aren't used in messenger
-  // generic arguments.
+
+  // Reject qualified-name type references, as we can't follow those.
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:438]: I removed the `istanbul ignore` comment because we can, in fact, write a test for this.

File: packages/platform-api-docs/src/extraction.ts [hunk 7]
```diff
@@ -322,268 +272,385 @@ function buildMethodInfo(method: MethodDeclaration): MethodInfo {
...
+        result.capabilityTypeDeclarations.push(
+          ...innerResult.capabilityTypeDeclarations,
+        );
+        for (const typeDeclaration of innerResult.visitedTypeDeclarations) {
+          result.visitedTypeDeclarations.add(typeDeclaration);
+        }
         continue;
       }
+
+      // TODO: Is this necessary?
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:496]: I am not sure that this logic makes sense to me, but I've kept it in for now. I will remove it in another PR.

File: packages/platform-api-docs/src/extraction.ts [hunk 8]
```diff
@@ -322,268 +272,385 @@ function buildMethodInfo(method: MethodDeclaration): MethodInfo {
...
+  capabilityTypeDeclaration: MessengerCapabilityTypeDeclaration,
   projectPath: string,
-): ExtractedMessengerCapabilityType | null {
-  const inlineCapability = extractFromInlineMessengerCapabilityType(
-    statement,
-    kind,
-    projectPath,
+): MessengerCapabilityPacket | null {
+  return (
+    tryToExtractFromMessengerCapabilityTypeLiteral(
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:552]: Simplified this and removed the `istanbul ignore` comment in the process.

File: packages/platform-api-docs/src/extraction.ts [hunk 9]
```diff
@@ -322,268 +272,385 @@ function buildMethodInfo(method: MethodDeclaration): MethodInfo {
...
+ * messenger action or event, extracted in a previous step.
  * @param projectPath - Project root, used for computing relative source paths.
- * @returns The extracted capability, or null if the shape doesn't match.
+ * @returns The extracted capability packet, or null if the shape of the type
+ * doesn't match.
  */
-function extractFromInlineMessengerCapabilityType(
-  statement: TypeAliasDeclaration | InterfaceDeclaration,
-  kind: 'action' | 'event',
+function tryToExtractFromMessengerCapabilityTypeLiteral(
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:579]: Renamed to `tryTo*` to signify that this is a part of a function where multiple strategies are being attempted.

File: packages/platform-api-docs/src/extraction.ts [hunk 10]
```diff
@@ -322,268 +272,385 @@ function buildMethodInfo(method: MethodDeclaration): MethodInfo {
...
+  //     handler: FooController['getState'];
+  //   }
+  //   // Bad
+  //   type FooControllerSomeAction = {};
+  //   // Bad
+  //   interface FooControllerSomeAction {};
   let members: TypeElementTypes[] | undefined;
-  if (NodeGuards.isTypeAliasDeclaration(statement)) {
-    const body = statement.getTypeNode();
+  if (NodeGuards.isTypeAliasDeclaration(declaration)) {
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:602]: Previously we were using `statement` to represent a type declaration, but I felt that was confusing. I guess a type declaration is a statement, but we've been using "declaration" pretty consistent elsewhere in this file.

File: packages/platform-api-docs/src/extraction.ts [hunk 11]
```diff
@@ -322,268 +272,385 @@ function buildMethodInfo(method: MethodDeclaration): MethodInfo {
...
+
+  let { description: jsDoc, params, returns } = extractJsDoc(declaration);
 
-  // For actions, if the handler resolves to a class method (`Class['method']`
-  // — possibly in another file), inherit its signature plus any JSDoc fields
-  // the type alias itself doesn't already provide.
+  // For actions, if the handler resolves to a class method (e.g.
+  // `FooController['method']`), inherit its signature plus any JSDoc fields the
+  // type alias itself doesn't already provide.
+  // TODO: We don't want to do this.
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:646]: I also questioned this logic because we ought to be able to rely on the stated JSDoc for an action or event type instead of needing to fill it in. But again I have left this here for now and will remove this in another PR.

File: packages/platform-api-docs/src/extraction.ts [hunk 12]
```diff
@@ -596,167 +663,209 @@ function extractFromInlineMessengerCapabilityType(
...
-      // istanbul ignore next: numeric/boolean literal types aren't valid as a
-      // messenger `type` and don't appear in real fixtures.
-      return null;
-    }
-    if (NodeGuards.isTemplateLiteralTypeNode(typeNode)) {
-      return resolveTemplateLiteralType(typeNode);
+function getMessengerCapabilityTypeString(
+  capabilityTypeProperties: PropertySignature[],
+): string | null {
+  const typeProperty = findProperty(capabilityTypeProperties, 'type');
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:694]: Extracted the `isPropertySignature` and `isIdentifier` checks to `findProperty`.

File: packages/platform-api-docs/src/extraction.ts [hunk 13]
```diff
@@ -596,167 +663,209 @@ function extractFromInlineMessengerCapabilityType(
...
+  //   }
+  //   type FooControllerSomeAction = {
+  //     type: `${typeof CONTROLLER_NAME}:someAction`;
+  //           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  //   }
+  const resolvedType = typeNode.getType();
+  if (resolvedType.isStringLiteral()) {
+    // Type assertion: There aren't any type guards we can use to narrow this
+    // type further.
+    const literalValue = resolvedType.getLiteralValueOrThrow() as string;
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:723]: As elsewhere, we don't need to use multiple strategies to resolve the string literal, we can just ask the type checker to resolve it for us.

File: packages/platform-api-docs/src/extraction.ts [hunk 14]
```diff
@@ -596,167 +663,209 @@ function extractFromInlineMessengerCapabilityType(
...
-    if (!NodeGuards.isPropertySignature(member)) {
+  for (const property of propertySignatures) {
+    if (!NodeGuards.isPropertySignature(property)) {
       continue;
     }
-    const memberNameNode = member.getNameNode();
+
+    const propertyNameNode = property.getNameNode();
     if (
-      // istanbul ignore next: capability properties always have identifier names.
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:686]: We don't need to make this assumption, we can check for it and exercise this logic in a test.

File: packages/platform-api-docs/src/extraction.ts [hunk 15]
```diff
@@ -596,167 +663,209 @@ function extractFromInlineMessengerCapabilityType(
...
-      return resolveTemplateLiteralType(typeNode);
+function getMessengerCapabilityTypeString(
+  capabilityTypeProperties: PropertySignature[],
+): string | null {
+  const typeProperty = findProperty(capabilityTypeProperties, 'type');
+  if (!typeProperty) {
+    return null;
+  }
+
+  // A `type` property without an explicit type wouldn't compile.
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:699]: Changed the `istanbul ignore` to a non-null assertion to better highlight that we are making an assumption.

File: packages/platform-api-docs/src/extraction.ts [hunk 16]
```diff
@@ -596,167 +663,209 @@ function extractFromInlineMessengerCapabilityType(
...
+
+  // We want our type alias to be a reference to a utility type.
+  // Non-null assertion: We can assume the type has a body of some kind,
+  // otherwise it wouldn't compile.
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  const body = declaration.getTypeNode()!;
+  // We already determined in
+  // `recursivelyFindMessengerCapabilityTypeDeclarations` that the declaration
+  // is a type reference.
+  // TODO: Capture the following information ahead of time so we don't need to
```
Comment (mcmire) [inline packages/platform-api-docs/src/extraction.ts:795]: There are more refactors we can do to simplify this function, but I decided to stop short of that so that this PR didn't blow up.

File: packages/platform-api-docs/src/types.ts [hunk 17]
```diff
@@ -8,10 +8,10 @@ export type DocumentedParameter = {
 };
 
 /**
- * One messenger capability — an action or event registered with the platform
- * — distilled from its TypeScript definition.
+ * Information about a messenger action or event extracted from its type
+ * in a source file.
  */
-export type ExtractedMessengerCapabilityType = {
+export type MessengerCapabilityPacket = {
```
Comment (mcmire) [inline packages/platform-api-docs/src/types.ts:14]: While it's true the information captured here does come *from* a capability type, I felt that saying that it *is* a type was confusing. Plus, we are using "item" elsewhere, so it didn't feel like we had a good term we could use consistently for this concept. I came up with "packet" as I felt like it was a good word to represent the bundle of information we've collected for a capability.

Note that I didn't update the use of "item" across the package. I only updated this type. I will make more changes in another PR to address that inconsistency.

**Review by mcmire — COMMENTED** [review]
Hey @cryptodev-2s, I hope you don't mind that I made some changes to `extraction.ts` as I went through it and tried to understand it. Hopefully these make sense but let me know if you have any questions.
